### PR TITLE
feat: create expiring threshold

### LINKED ISSUE #89: What happens if the access token is about to expire?
The AbaxAuth class will refresh an expired access token (if able) when you call `AbaxAuth.getAccessToken`. It will not refresh the access token if the token is about to expire, say in 1 second, 30 seconds, 1 minute or 5 minutes. 

It's quite plausible that whatever the access token is subsequently used for will face a 401 error if the AbaxAuth naïvely tells you "this token is good to go 👍" when it is in fact expiring very shortly. For example, a program fetching thousands of entities that needs to slow down due to rate limiting could easily last more than a few minutes. If you know that your program falls into this category, it would be nice to be able to tell AbaxAuth that you won't accept an access token that is about to expire very shortly.

Perhaps we should add an option to the AbaxAuth class or getAccessToken function that lets the user specify an amount of time that the token ought to last.

That could look like this:

```ts
const abaxAuth = new AbaxAuth({ 
  clientId: 'abc', 
  clientSecret: 'def', 
  redirectUri: 'http://localhost:3000', 
});

abaxAuth.setCredentials({
  accessToken: 'existingToken',
  refreshToken: 'refreshToken',
  expiresAt: new Date('2024-04-10T12:00'),
  tokenType: 'Bearer',
});

const apiKey = abaxAuth.getAccessToken(600); // get an access token that must last at least 600 seconds from now
```

Alternatively, it could look like this:

```ts

const abaxAuth = new AbaxAuth({ 
  clientId: 'abc', 
  clientSecret: 'def', 
  redirectUri: 'http://localhost:3000', 
  minimumAccessTokenDuration: 600, // access tokens must last at least 600 seconds when retrieved
});

abaxAuth.setCredentials({
  accessToken: 'existingToken',
  refreshToken: 'refreshToken',
  expiresAt: new Date('2024-04-10T12:00'),
  tokenType: 'Bearer',
});

const apiKey = abaxAuth.getAccessToken(); 
```

### PR DESCRIPTION
**Pull request by Xillians**
Fixes #89

### CODE REVIEW

File: src/authentication/abax-auth.ts [hunk 1]
```diff
@@ -51,16 +51,27 @@ export interface AbaxAuthConfig {
    * @default https://identity.abax.cloud
    */
   baseUrl?: string | undefined;
+  /**
+   * The minimum lifetime of the token in seconds. If the token expires in less
+   * than this time, it will be refreshed.
+   */
+  minimumTokenLifetime?: number | undefined;
```
Comment (braaar) [inline src/authentication/abax-auth.ts:58]:
```suggestion
  /**
   * The minimum required lifetime of an access token in seconds. When getting an access token, the token will be 
   * refreshed accordingly to ensure the access token returned satisfies this requirement.
   */
  minimumTokenLifetime?: number | undefined;
```

File: src/authentication/abax-auth.ts [hunk 2]
```diff
@@ -51,16 +51,27 @@ export interface AbaxAuthConfig {
...
 }
 
 export class AbaxAuth {
   private baseUrl = 'https://identity.abax.cloud';
   private credentials?: AbaxCredentials;
+  private minimumTokenLifetime: number;
 
   constructor(private readonly config: AbaxAuthConfig) {
```
Comment (braaar) [inline src/authentication/abax-auth.ts:66]:
```suggestion
  constructor(private readonly {baseUrl, minimumTokenLifetime = 60 }: AbaxAuthConfig) {
```

Could something like this work? Feels awkward to have an if statement when we are just applying a default value. 

We should also prevent the user from setting a negative value and breaking things.

File: src/authentication/abax-auth.ts [hunk 3]
```diff
@@ -207,6 +218,14 @@ export class AbaxAuth {
...
+    if (
+      this.credentials.expiresAt.getTime() - Date.now() <
+      this.minimumTokenLifetime * 1000
+    ) {
+      await this.refreshCredentials();
+    }
+
     // check if token is expired
     if (
       this.credentials.refreshToken &&
```
Comment (braaar) [inline src/authentication/abax-auth.ts:231]: I think the extra step where we check if the token is expired is unnecessary here, so just remove it.

**Review by braaar — APPROVED** [review]
Looks good!

File: src/authentication/abax-auth.ts (hunk 2 above)
Comment (Xillians) [reply src/authentication/abax-auth.ts:66]: Updated to use `??` operator, as the `{variables}` in constructor was not a valid with error:
```
A parameter property may not be declared using a binding pattern.ts(1187)
```

**Pull request merged**